pkg/controllermanager/controller/managedseedset/add.go

```diff
@@ -0,0 +1,372 @@
+	managedSeed, ok := e.ObjectNew.(*seedmanagementv1alpha1.ManagedSeed)
+	if !ok {
+		return false
+	}
+
+	oldManagedSeed, ok := e.ObjectOld.(*seedmanagementv1alpha1.ManagedSeed)
+	if !ok {
+		return false
+	}
+
+	if !reflect.DeepEqual(managedSeed.DeletionTimestamp, oldManagedSeed.DeletionTimestamp) {
+		return true
+	}
+
+	return p.filterManagedSeed(e.ObjectNew, false)
+}
+
+func (p *managedSeedPredicate) Delete(e event.DeleteEvent) bool {
+	return p.filterManagedSeed(e.Object, true)
+}
+
+func (p *managedSeedPredicate) Generic(_ event.GenericEvent) bool { return false }
+
+// Returns true only if the ManagedSeed belongs to the pending replica and it progressed from the state
+// that caused the replica to be pending.
+func (p *managedSeedPredicate) filterManagedSeed(obj client.Object, deleted bool) bool {
+	managedSeed, ok := obj.(*seedmanagementv1alpha1.ManagedSeed)
+	if !ok {
+		return false
+	}
+
+	controllerRef := metav1.GetControllerOf(managedSeed)
+	if controllerRef == nil {
+		return false
+	}
+
+	managedSeedSet := &seedmanagementv1alpha1.ManagedSeedSet{}
+	if err := p.reader.Get(p.ctx, kutil.Key(managedSeed.Namespace, controllerRef.Name), managedSeedSet); err != nil {
+		return false
+	}
+
+	if managedSeedSet.Status.PendingReplica == nil || managedSeedSet.Status.PendingReplica.Name != managedSeed.Name {
+		return false
+	}
+
+	switch managedSeedSet.Status.PendingReplica.Reason {
+	case seedmanagementv1alpha1.ManagedSeedPreparingReason:
+		return managedSeedRegistered(managedSeed) || managedSeed.DeletionTimestamp != nil
+	case seedmanagementv1alpha1.ManagedSeedDeletingReason:
+		return deleted
+	default:
+		return false
+	}
+}
+
+// SeedPredicate returns the predicate for Seed events.
+func (r *Reconciler) SeedPredicate() predicate.Predicate {
+	return &seedPredicate{}
+}
+
+type seedPredicate struct {
+	ctx    context.Context
+	reader client.Reader
+}
+
+func (p *seedPredicate) InjectStopChannel(stopChan <-chan struct{}) error {
+	p.ctx = contextutil.FromStopChannel(stopChan)
+	return nil
+}
+
+func (p *seedPredicate) InjectClient(client client.Client) error {
+	p.reader = client
+	return nil
+}
+
+func (p *seedPredicate) Create(e event.CreateEvent) bool { return p.filterSeed(e.Object, false) }
+
+func (p *seedPredicate) Update(e event.UpdateEvent) bool {
+	seed, ok := e.ObjectNew.(*gardencorev1beta1.Seed)
+	if !ok {
+		return false
+	}
+
+	oldSeed, ok := e.ObjectOld.(*gardencorev1beta1.Seed)
+	if !ok {
+		return false
+	}
+
+	if seedReady(seed) != seedReady(oldSeed) {
+		return true
+	}
+
+	return p.filterSeed(e.ObjectNew, false)
+}
+
+func (p *seedPredicate) Delete(e event.DeleteEvent) bool { return p.filterSeed(e.Object, true) }
+
+func (p *seedPredicate) Generic(_ event.GenericEvent) bool { return false }
+
+// Returns true only if the Seed belongs to the pending replica and it progressed from the state
+// that caused the replica to be pending.
+func (p *seedPredicate) filterSeed(obj client.Object, _ bool) bool {
```

Comment:
same here (in this case it's even more useless since you don't even evaluate/use the second parameter)